### PR TITLE
Ui server ejs

### PR DESCRIPTION
**Pull request by Arman199116**

### CODE REVIEW

File: web/views/bteq.ejs [hunk 1]
```diff
@@ -1 +1,43 @@
...
+  <head>
+    <meta charset="utf-8" />
+    <link rel="stylesheet" type="text/css" href="/bteq.css" />
+
+    <title>BTEQ</title>
+  </head>
+  <body>
+    <div class="grid-container">
+        <header class="grid-item page_header">
+            Logon and Set Width commands
```
Comment (maryantonyan) [inline web/views/bteq.ejs]: "LOGON and SET WIDTH commands"

File: web/views/bteq.ejs [hunk 2]
```diff
@@ -1 +1,43 @@
...
+        </header>
+
+        <form method="POST" class="form_lable" action="http://localhost:3000">
+            Logon: <br />
```
Comment (maryantonyan) [inline web/views/bteq.ejs]: use `<label>` tag, provide styles for that, avoid using `<br/>` tags

File: web/views/bteq.ejs [hunk 3]
```diff
@@ -1 +1,43 @@
...
+            <input type="text" name="logon" placeholder=".logon host:port/user,pass"
```
Comment (maryantonyan) [inline web/views/bteq.ejs]: Provide styles instead of using `<br/>`

File: web/public/bteq.css [hunk 4]
```diff
@@ -1 +1,71 @@
...
+    grid-gap: 10px;
+    background-color: rgb(245, 245, 245);
+    padding: 10px;
+    height: 100%;
+}
+
+.grid-item {
+    background-color: #ecebeb;
+    text-align: center;
+    font-size: 20px;
```
Comment (maryantonyan) [inline web/public/bteq.css]: I guess this property has no effect anywhere, please check.
image

File: web/views/bteq.ejs [hunk 5]
```diff
@@ -1 +1,43 @@
...
+            Select: <br />
+            <input type="text" name="select" placeholder="select..." /> <br />
+            <input type="submit" value="Submit" class="submit_class" />
+        </form>
+        <div class="grid-item output_div">
+            <pre> <%= data %> </pre>
+        </div>
+
+        <footer class="grid-item page_footer">
+            BTEQ is optimized for learning, testing, and training.
```
Comment (maryantonyan) [inline web/views/bteq.ejs]: Please, update the footer to describe the implemented tool and BTEQ itself.

> An interface that works in place of the Teradata BTEQ interface. BTEQ is a general-purpose, command-based program that allows users to communicate with one or more Teradata servers, and to format reports for both print and screen output. BTEQ allows users to submit SQL queries to Teradata Database. BTEQ formats the results and returns them to the screen, a file, or to a printer.

File: web/public/bteq.css [hunk 6]
```diff
@@ -1 +1,71 @@
...
+    grid-row:  2 / 3;
+    text-align: left;
+    font-size: 15px;
+    padding: 20px;
+    overflow-y: scroll;
+    word-wrap: break-word;
+    color: #f3753f;
+}
+
+.form_lable {
```
Comment (maryantonyan) [inline web/public/bteq.css]: Minor: lable -> label

File: web/views/bteq.ejs (hunk 2 above)
Comment (maryantonyan) [reply web/views/bteq.ejs]: Also, please add a little bottom margin to the label.

File: web/public/bteq.css [hunk 7]
```diff
@@ -1 +1,71 @@
...
+
+.form_lable > input {
+    border-radius: 13px;
+    height: 22px;
+    width: 240px;
+    padding-left: 6px;
+    margin-bottom: 25px;
+}
+
+.form_lable > .submit_class {
```
Comment (maryantonyan) [inline web/public/bteq.css]: Please, align submit button to the right (with red line):
image

File: web/server.js [hunk 8]
```diff
@@ -1 +1,35 @@
...
+
+var cmd_file = 'cmd.txt';
+const port = 3000;
+
+app.get('/', (req, res) => {
+    res.render('bteq', {data : ""});
+});
+
+app.post('/', (req, res) => {
+
```
Comment (maryantonyan) [inline web/server.js:24]: remove empty line

File: web/server.js [hunk 9]
```diff
@@ -1 +1,35 @@
...
+var express = require('express');
+var app = express();
+var bodyparser = require('body-parser');
+var fs = require('fs');
+var open = require('open');
+var exec_cmd = require('./exec_cmd');
+
+app.set('view engine', 'ejs');
+app.use(express.static('public'));
+app.use(bodyparser.urlencoded({ extended : false }));
```
Comment (maryantonyan) [inline web/server.js]: remove spaces:
`app.use(bodyparser.urlencoded({extended : false}));`

File: web/server.js [hunk 10]
```diff
@@ -1 +1,35 @@
...
+const port = 3000;
+
+app.get('/', (req, res) => {
+    res.render('bteq', {data : ""});
+});
+
+app.post('/', (req, res) => {
+
+    var {'logon' : log, 'set-width' : wid, 'select' : sel} = req.body;
+    fs.writeFile(cmd_file, `${log}\n${wid}\n${sel}`, async function (err) {
```
Comment (maryantonyan) [inline web/server.js]: remove the space:
`fs.writeFile(cmd_file, `${log}\n${wid}\n${sel}`, async function(err) {`

File: web/server.js [hunk 11]
```diff
@@ -1 +1,35 @@
...
+        if (err) throw err;
+        await exec_cmd(cmd_file);
+        fs.readFile('out.bteq', 'utf8', function(err, data) {
+            if (err) throw err;
+            res.render('bteq', {data} );
```
Comment (maryantonyan) [inline web/server.js]: res.render('bteq', {data});

File: web/views/bteq.ejs [hunk 12]
```diff
@@ -1 +1,43 @@
...
+            constantly reviewed to avoid errors, but we cannot warrant
+            full correctness of all content. While using this site, you
+            agree to have read and accepted our terms of use, cookie and
+            privacy policy. Copyright 1999-2019 by Refsnes
+            Data. All Rights Reserved.
+        </footer>
+    </div>
+  </body>
+</html>
+
```
Comment (maryantonyan) [inline web/views/bteq.ejs]: remove empty line

File: web/exec_cmd.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
+const util = require('util');
+const exec = util.promisify(require('child_process').exec);
+var fs = require('fs');
+
+var bteq_path = `./bteq_dir/psql -m 1 < ${__dirname}/`;
+var out_file = ` > ${__dirname}/out.bteq`;
+
+module.exports = async (cmd_file) => {
```
Comment (maryantonyan) [inline web/exec_cmd.js]: Remove the space:
`module.exports = async(cmd_file) => {`

File: web/server.js [hunk 14]
```diff
@@ -1 +1,35 @@
...
+        });
+    });
+});
+
+app.listen(port, function () {
+    console.log("Example app listening at http://127.0.0.1:%s", port);
```
Comment (maryantonyan) [inline web/server.js]: `Example app`??

**Review by maryantonyan — CHANGES_REQUESTED** [review]
Here are my comments for now, later will through the .js codes one more time.

File: web/exec_cmd.js [hunk 15]
```diff
@@ -0,0 +1,20 @@
...
+var bteq_path = `./bteq_dir/psql -m 1 < ${__dirname}/`;
+var out_file = ` > ${__dirname}/out.bteq`;
+
+module.exports = async(cmd_file) => {
+    try {
+        const {stdout, stderr} = await exec(`${bteq_path}${cmd_file}${out_file}`);
+        console.log('stdout:', stdout);
+        try {
+            fs.unlinkSync(cmd_file);
+        } catch(err) {
```
Comment (M7Arman) [inline web/exec_cmd.js]: you can miss this try/catch, the parent try/catch will catch any exception of unlinkSync function

File: web/server.js [hunk 16]
```diff
@@ -1 +1,51 @@
...
+var fs = require('fs');
+var open = require('open');
+var exec_cmd = require('./exec_cmd');
+
+app.set('view engine', 'ejs');
+app.use(express.static('public'));
+app.use(bodyparser.urlencoded({extended : false}));
+
+var cmd_file = 'cmd.txt';
+const port = 3000;
```
Comment (M7Arman) [inline web/server.js:13]: Why cmd_file is defined using **var** keyword and port is defined with **const** keyword?

File: web/server.js [hunk 17]
```diff
@@ -1 +1,51 @@
...
+
+app.get('/', (req, res) => {
+    res.render('bteq', {data : ""});
+});
+
+app.post('/', (req, res) => {
+    var {'logon' : log, 'set-width' : wid, 'select' : sel} = req.body;
+    var logon = undefined;
```
Comment (M7Arman) [inline web/server.js]: can you use **let** instead of **var**? Also, please read why it's more preferable.

File: web/server.js [hunk 18]
```diff
@@ -1 +1,51 @@
...
+    var set = undefined;
+    var select = undefined;
+
```
Comment (M7Arman) [inline web/server.js:24]: why did you initialize the variables with undefined?

File: web/server.js [hunk 19]
```diff
@@ -1 +1,51 @@
...
+        set = wid.match(/^\.[a-zA-Z0-9_$]/g);
+    }
+    if (!!sel.trim()) {
+        select = sel.match(/^select/gi);
+    }
+    if (logon === null || set === null || select === null) {
+        res.render('bteq', {data : "Incorrect input value"});
+    } else {
+        fs.writeFile(cmd_file, `${log}\n${wid}\n${sel}`, async function(err) {
+            if (err) throw err;
```
Comment (M7Arman) [inline web/server.js]: no need to catch this exception?

File: web/server.js [hunk 20]
```diff
@@ -1 +1,51 @@
...
+            await exec_cmd(cmd_file);
+            fs.readFile('out.bteq', 'utf8', function(err, data) {
+                if (err) throw err;
```
Comment (M7Arman) [inline web/server.js]: no need to catch this exception?

File: web/server.js [hunk 21]
```diff
@@ -1 +1,59 @@
...
+            fs.writeFile(cmd_file, `${log}\n${wid}\n${sel}`, async(err) => {
+                if (err) throw err;
+                await exec_cmd(cmd_file);
+                try {
+                    fs.readFile('out.bteq', 'utf8', (err, data) => {
+                        if (err) throw err;
+                        res.render('bteq', {data});
+                    });
+                } catch(err) {
+                    if (err) console.error(err);
```
Comment (M7Arman) [inline web/server.js:52]: in API functions you HAVE TO respond to the client. With your implementation, you didn't respond in this case.

File: web/server.js [hunk 22]
```diff
@@ -1 +1,59 @@
...
+                }
+            });
+        } catch(err) {
+            if (err) console.error(err);
```
Comment (M7Arman) [inline web/server.js:56]: the same here. missing response handling

**Review by maryantonyan — APPROVED** [review]

**Pull request merged**